### PR TITLE
ENH: Add JavaScript UDFs for BigQuery

### PR DESCRIPTION
**Pull request by cpcloud**

### CODE REVIEW

File: ibis/bigquery/udf/udf.py [hunk 1]
```diff
@@ -0,0 +1,820 @@
...
+    -------
+    indented_lines : str
+    """
+    if isinstance(lines, six.string_types):
+        text = [lines]
+    text = '\n'.join(lines)
+    return textwrap.indent(text, ' ' * spaces)
+
+
+def udf(input_type, output_type, strict=True):
```
Comment (kszucs) [inline ibis/bigquery/udf/udf.py]: How about moving this to `ibis.bigquery.udf.api` or something?

File: ibis/bigquery/udf/udf.py [hunk 2]
```diff
@@ -0,0 +1,820 @@
...
+            value=node.func.value,
+            attr='push',
+            ctx=ast.Load(),
+        ),
+        args=node.args,
+        keywords=node.keywords,
+    )
+
+
+class PythonToJavaScriptTranslator:
```
Comment (kszucs) [inline ibis/bigquery/udf/udf.py]: How about moving this to `transpiler.py` along with the `find` utility in `find.py`? Find is only used during translation (maybe `Rewriter` too).

File: ibis/client.py [hunk 3]
```diff
@@ -221,15 +218,12 @@ def execute(self, expr, params=None, limit='default', async=False,
           Scalar expressions: Python scalar value
         """
         ast = self._build_ast_ensure_limit(expr, limit, params=params)
+        result = self._execute_query(ast, async=async, **kwargs)
```
Comment (kszucs) [inline ibis/client.py]: `ast` got confusing with js2py transpiler introduced, we might prefer `query_ast` here.

File: ibis/sql/alchemy.py [hunk 4]
```diff
@@ -683,6 +670,18 @@ def __init__(self, *args, **kwargs):
         super(AlchemyContext, self).__init__(*args, **kwargs)
         self._table_objects = {}
 
+    def collapse(self, queries):
+        if not isinstance(queries, six.string_types):
```
Comment (kszucs) [inline ibis/sql/alchemy.py]: Maybe the following is more idiomatic:

```py
if isinstance(queries, six.string_types):
    return query

if len(queries) > 1:
    raise 'Multi-statement queries not supported'

return queries[0]
```

**Comment by kszucs** [comment]
Time's up for modern frontend dialects!

image

File: ibis/bigquery/compiler.py [hunk 5]
```diff
@@ -8,25 +10,49 @@
...
+        self.expr = expr
+        self.context = context
+
+    def compile(self):
+        return self.expr.op().js
+
+
+def find_bigquery_udf(expr):
+    if isinstance(expr.op(), BigQueryUDFNode):
+        return lin.halt, expr
```
[the remark this answers is not in the document]
Comment (cpcloud) [reply ibis/bigquery/compiler.py]: Nope. Fixing.

[the remark this answers is not in the document]
Comment (cpcloud) [reply ibis/bigquery/compiler.py]: Fixed. Multiple UDFs now tested and working

File: ibis/bigquery/compiler.py [hunk 6]
```diff
@@ -8,25 +10,49 @@
...
     select_builder = BigQuerySelectBuilder
 
-    @property
-    def _union_class(self):
-        # return BigQueryUnion
-        raise NotImplementedError()
+    def generate_setup_queries(self):
+        result = list(
+            map(partial(BigQueryUDFDefinition, context=self.context),
+                lin.traverse(find_bigquery_udf, self.expr)))
```
[the remark this answers is not in the document]
Comment (cpcloud) [reply ibis/bigquery/compiler.py]: it might :)

File: ibis/bigquery/udf/__init__.py [hunk 7]
```diff
@@ -0,0 +1 @@
+from ibis.bigquery.udf.udf import udf  # noqa: F401
```
[the remark this answers is not in the document]
Comment (cpcloud) [reply ibis/bigquery/udf/__init__.py]: Ok, I'll rename the udf module to `core`.

File: ibis/bigquery/udf/udf.py (hunk 1 above)
Comment (cpcloud) [reply ibis/bigquery/udf/udf.py]: done

File: ibis/client.py (hunk 3 above)
Comment (cpcloud) [reply ibis/client.py]: renamed.

File: ibis/sql/alchemy.py (hunk 4 above)
Comment (cpcloud) [reply ibis/sql/alchemy.py]: yep, agree.

Comment (cpcloud) [reply ibis/sql/alchemy.py]: done

File: ibis/bigquery/compiler.py [hunk 8]
```diff
@@ -36,8 +62,8 @@ def build_ast(expr, context):
 
 def _get_query(expr, context):
     ast = build_ast(expr, context)
-    (query, rest) = (ast.queries[0], ast.queries[1:])
-    assert not rest
+    query, rest = (ast.queries[0], ast.queries[1:])
+    assert not rest, '*rest should be empty in bigquery.compiler._get_query'
```
[the remark this answers is not in the document]
Comment (cpcloud) [reply ibis/bigquery/compiler.py]: I'm removing this function from this module.

File: ibis/bigquery/compiler.py [hunk 9]
```diff
@@ -8,42 +10,61 @@
...
+        self.context = context
+
+    def compile(self):
+        return self.expr.op().js
+
+
+def find_bigquery_udf(expr):
+    if isinstance(expr.op(), BigQueryUDFNode):
+        result = expr
+        return lin.proceed, expr
```
Comment (kszucs) [inline ibis/bigquery/compiler.py]: I guess this is not required.

File: ibis/bigquery/client.py [hunk 10]
```diff
@@ -291,10 +291,11 @@ def _build_ast(self, expr, context):
         result = comp.build_ast(expr, context)
         return result
 
-    def _execute_query(self, ddl, async=False):
+    def _execute_query(self, dml, async=False):
```
[the remark this answers is not in the document]
Comment (cpcloud) [reply ibis/bigquery/client.py]: Actually I think you're right. Let me take another look.

**Pull request closed without merging**